In [2]:
import pandas as pd

product_details = pd.read_excel('data.xlsx', sheet_name='product_details')
retail_data1    = pd.read_excel('data.xlsx', sheet_name='retail_data1')
retail_data2    = pd.read_excel('data.xlsx', sheet_name='retail_data2')



Initial Data Assessment

In [8]:
print(retail_data1.shape)
print(retail_data1.dtypes)
print(retail_data1.isnull().sum())
print(retail_data1.duplicated().sum())

(4243, 16)
transaction_id         int64
customer_id            int64
customer_name         object
product_id             int64
price                float64
product_name          object
category              object
purchase_location     object
city                  object
transaction_date      object
quantity               int64
payment_method        object
discount             float64
email                 object
phone                  int64
payment_status        object
dtype: object
transaction_id         0
customer_id            0
customer_name          0
product_id             0
price                404
product_name           0
category               0
purchase_location      0
city                   0
transaction_date       0
quantity               0
payment_method         0
discount               0
email                  0
phone                  0
payment_status         0
dtype: int64
0


Findings-Retail data1
- Dataset contains 4,243 records and 16 columns.
- Schema successfully loaded.
- transaction_date is stored as object datatype.
- 404 missing values found in price column.
- No duplicate records found.
- Further validation required for quantity, discount, category, and product consistency.

In [9]:
print(retail_data2.shape)
print(retail_data2.dtypes)
print(retail_data2.isnull().sum())
print(retail_data2.duplicated().sum())

(4251, 16)
transaction_id         int64
customer_id            int64
customer_name         object
product_id             int64
price                float64
product_name          object
category              object
purchase_location     object
city                  object
transaction_date      object
quantity               int64
payment_method        object
discount             float64
email                 object
phone                  int64
payment_status        object
dtype: object
transaction_id         0
customer_id            0
customer_name          0
product_id             0
price                405
product_name           0
category               0
purchase_location      0
city                   0
transaction_date       0
quantity               0
payment_method         0
discount               0
email                  0
phone                  0
payment_status         0
dtype: int64
0


Initial Assessment - retail_data2

Issues Identified
- Dataset contains 4,251 records and 16 columns.
- transaction_date is stored as object datatype and requires datetime conversion.
- 405 missing values found in the price column.
- Customer name, email, and phone contain PII data and require masking.
- No duplicate records detected.
- Further validation required for quantity, discount, category, and product consistency.
- transaction_date is currently stored as object datatype instead of datetime datatype.
- Date standardization and datetime conversion will be required before performing time-based analysis.

In [11]:
print(set(retail_data1.columns) == set(retail_data2.columns))
print(retail_data1.dtypes.equals(retail_data2.dtypes))

True
True


scehma verified both the tables have same columns and data types 

In [13]:
retail_data1.head()
#retail_data2.head()

,transaction_id,customer_id,customer_name,product_id,price,product_name,category,purchase_location,city,transaction_date,quantity,payment_method,discount,email,phone,payment_status
0,1,642,Troy Mitchell,108,15000.0,Mixer grinder,Home Appliances,offline,Bangalore,2025-12-06 00:00:00,3,UPI,0.10,Troy60@gmail.com,8385276968,successful
1,2,881,Sarah Guerrero,102,70000.0,Phone,ELEC,online,Bangalore,2025-07-21 00:00:00,5,Card,0.35,SarahGuerrero@yahoo.com,7147248911,successful
2,3,505,Samantha Hull,109,65000.0,Refrigerator,Home Appliances,offline,Chennai,2025-07-11 00:00:00,2,UPI,0.05,SamanthaHull@outlook.com,7415321565,successful
3,4,794,Gerald Cooper,106,80000.0,sofa,FURN,offline,Chennai,2025-10-06 00:00:00,2,UPI,0.05,GeraldCooper@gmail.com,9739354201,successful
4,5,864,Cameron Black,108,15000.0,Mixer grinder,home appliances,offline,Chennai,2025-12-12 00:00:00,1,Cash,0.05,CameronBlack@yahoo.com,9938169477,successful


### Columns Present

- transaction_id
- customer_id
- customer_name
- product_id
- price
- product_name
- category
- purchase_location

In [17]:
def profile_dataset(df, name):

    print(f"\n{'='*50}")
    print(f"DATASET: {name}")
    print(f"{'='*50}")

    print(f"Shape: {df.shape[0]} rows × {df.shape[1]} columns")

    print("\n--- Missing Values ---")
    nulls = df.isnull().sum()
    null_pct = (nulls / len(df) * 100).round(2)

    print(
        pd.DataFrame({
            'missing_count': nulls,
            'missing_%': null_pct
        })[nulls > 0]
    )

    print("\n--- Duplicates ---")
    print(f"Fully duplicate rows: {df.duplicated().sum()}")

    if 'transaction_id' in df.columns:
        print(
            f"Duplicate transaction_ids: "
            f"{df['transaction_id'].duplicated().sum()}"
        )

    print("\n--- Data Types ---")
    print(df.dtypes)

    print("\n--- Categorical Columns ---")

    categorical_cols = [
        'category',
        'product_name',
        'purchase_location',
        'payment_method',
        'payment_status'
    ]

    for col in categorical_cols:
        if col in df.columns:
            print(f"\n{col}")
            print(f"Unique Values: {df[col].nunique()}")
            print(df[col].unique()[:10])

    print("\n--- Numeric Summary ---")

    numeric_cols = [
        col for col in
        ['price', 'quantity', 'discount']
        if col in df.columns
    ]

    print(df[numeric_cols].describe())
    
profile_dataset(retail_data1, "Retail Data 1")


DATASET: Retail Data 1
Shape: 4243 rows × 16 columns

--- Missing Values ---
       missing_count  missing_%
price            404       9.52

--- Duplicates ---
Fully duplicate rows: 0
Duplicate transaction_ids: 243

--- Data Types ---
transaction_id         int64
customer_id            int64
customer_name         object
product_id             int64
price                float64
product_name          object
category              object
purchase_location     object
city                  object
transaction_date      object
quantity               int64
payment_method        object
discount             float64
email                 object
phone                  int64
payment_status        object
dtype: object

--- Categorical Columns ---

category
Unique Values: 12
['Home Appliances' 'ELEC' 'FURN' 'home appliances' 'Electronics'
 'electronics' 'CLOTH' 'clothing' 'Clothing' 'furniture']

product_name
Unique Values: 30
['Mixer grinder' 'Phone' 'Refrigerator' 'sofa' 'laptop' 'phone'
 'Microwa

## Profiling Findings - retail_data1

### Data Quality Issues Identified

- 404 missing values detected in the price column (9.52%).
- 243 duplicate transaction IDs detected.
- transaction_date is stored as object datatype and requires datetime conversion.
- Category values contain inconsistent naming conventions and casing.
- Product names contain inconsistent capitalization and formatting.
- Invalid quantity values detected (minimum quantity = -2).
- Customer information contains PII fields requiring masking.
- No fully duplicated rows were identified.

In [18]:
def profile_dataset(df, name):

    print(f"\n{'='*50}")
    print(f"DATASET: {name}")
    print(f"{'='*50}")

    print(f"Shape: {df.shape[0]} rows × {df.shape[1]} columns")

    print("\n--- Missing Values ---")
    nulls = df.isnull().sum()
    null_pct = (nulls / len(df) * 100).round(2)

    print(
        pd.DataFrame({
            'missing_count': nulls,
            'missing_%': null_pct
        })[nulls > 0]
    )

    print("\n--- Duplicates ---")
    print(f"Fully duplicate rows: {df.duplicated().sum()}")

    if 'transaction_id' in df.columns:
        print(
            f"Duplicate transaction_ids: "
            f"{df['transaction_id'].duplicated().sum()}"
        )

    print("\n--- Data Types ---")
    print(df.dtypes)

    print("\n--- Categorical Columns ---")

    categorical_cols = [
        'category',
        'product_name',
        'purchase_location',
        'payment_method',
        'payment_status'
    ]

    for col in categorical_cols:
        if col in df.columns:
            print(f"\n{col}")
            print(f"Unique Values: {df[col].nunique()}")
            print(df[col].unique()[:10])

    print("\n--- Numeric Summary ---")

    numeric_cols = [
        col for col in
        ['price', 'quantity', 'discount']
        if col in df.columns
    ]

    print(df[numeric_cols].describe())
    
profile_dataset(retail_data1, "Retail Data 2")


DATASET: Retail Data 2
Shape: 4243 rows × 16 columns

--- Missing Values ---
       missing_count  missing_%
price            404       9.52

--- Duplicates ---
Fully duplicate rows: 0
Duplicate transaction_ids: 243

--- Data Types ---
transaction_id         int64
customer_id            int64
customer_name         object
product_id             int64
price                float64
product_name          object
category              object
purchase_location     object
city                  object
transaction_date      object
quantity               int64
payment_method        object
discount             float64
email                 object
phone                  int64
payment_status        object
dtype: object

--- Categorical Columns ---

category
Unique Values: 12
['Home Appliances' 'ELEC' 'FURN' 'home appliances' 'Electronics'
 'electronics' 'CLOTH' 'clothing' 'Clothing' 'furniture']

product_name
Unique Values: 30
['Mixer grinder' 'Phone' 'Refrigerator' 'sofa' 'laptop' 'phone'
 'Microwa

## Profiling Findings - Retail Data 2

### Data Quality Issues Identified

- Dataset contains 4,243 records and 16 columns.
- 404 missing values detected in the price column (9.52% of total records).
- 243 duplicate transaction IDs identified.
- transaction_date is stored as object datatype and requires conversion to datetime format.
- Category values contain inconsistent naming conventions and capitalization (e.g., Electronics, electronics, ELEC).
- Product names contain inconsistent capitalization and formatting (e.g., Phone, phone, PHONE).
- Invalid quantity values detected (minimum quantity = -2).
- Customer information contains Personally Identifiable Information (PII) including customer_name, email, and phone.
- No fully duplicated rows were identified.
- Payment method and payment status values appear standardized and require no immediate correction.

## Data Cleaning Plan

| Issue | Action |
|---------|---------|
| Missing prices | Recover using product dimension table |
| Duplicate transaction IDs | Investigate and remove duplicate transactions |
| Category inconsistencies | Standardize category names |
| Product name inconsistencies | Standardize product names |
| Negative quantities | Remove invalid records |
| transaction_date datatype | Convert to datetime |
| PII fields | Mask email and phone |

In [3]:
category_mapping = {
'ELEC': 'Electronics',
'electronics': 'Electronics',
'CLOTH': 'Clothing',
'clothing': 'Clothing',
'FURN': 'Furniture',
'furniture': 'Furniture',
'home appliances': 'Home Appliances',
'HOME': 'Home Appliances'
}

retail_data1['category'] = retail_data1['category'].replace(category_mapping)
retail_data2['category'] = retail_data2['category'].replace(category_mapping)

In [5]:
retail_data2['category'].unique()

array(['Furniture', 'Electronics', 'Home Appliances', 'Clothing'],
      dtype=object)

In [ ]:
#retail_data1['product_name'] = retail_data1['product_name'].str.title()
retail_data2['product_name'] = retail_data2['product_name'].str.title()

In [10]:
retail_data1['product_name'].unique()

array(['Mixer Grinder', 'Phone', 'Refrigerator', 'Sofa', 'Laptop',
       'Microwave', 'Shoes', 'Shirt', 'Tv', 'Dining Table'], dtype=object)

In [11]:
retail_data1['transaction_date'] = pd.to_datetime(
retail_data1['transaction_date'],
errors='coerce'
)

#retail_data2['transaction_date'] = pd.to_datetime(
#retail_data2['transaction_date'],
#errors='coerce'
#)